Worker daemon not shutting down sub agents - causing massive API spend

- worker-service.cjs --daemon spawns claude --model claude-sonnet-4-5     
  subprocesses that never exit                                              
  - ~30+ accumulated over ~36 hours (Wed 10PM → Thu 11AM)                   
  - Each had 5-12+ min CPU time, all hitting the Anthropic API continuously 
  - The --resume flag on many suggests they're retrying/resuming the same   
  sessions repeatedly   

### Environment
- **Claude-mem version**:
- **Claude Code version**:
- **OS**:
- **Platform**:

### Logs
╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌
    1 # claude-mem: Runaway subagent processes causing excessive API spen   
      d                                                                     
    2
    3 **Plugin**: claude-mem v9.0.12 (thedotmack)                           
    4 **Date**: 2026-04-03
    5 **Impact**: ~$54+ in Anthropic API charges from zombie `claude-sonn   
      et-4-5` processes                                                     
    6
    7 ## Summary
    8
    9 The claude-mem worker daemon spawns `claude` CLI subprocesses (one
      per observed session) that never terminate. These accumulate across
       sessions and persist indefinitely, continuously consuming API toke
      ns. After ~36 hours of normal usage, 30+ background processes were
      running simultaneously, all calling `claude-sonnet-4-5`.
   10
   11 ## Configuration
   12
   13 ```json
   14 // ~/.claude-mem/settings.json
   15 {
   16   "CLAUDE_MEM_MODEL": "claude-sonnet-4-5",
   17   "CLAUDE_MEM_PROVIDER": "claude",
   18   "CLAUDE_MEM_WORKER_PORT": "37777",
   19   "CLAUDE_MEM_WORKER_HOST": "127.0.0.1"
   20 }
   21 ```
   22
   23 ## Evidence
   24
   25 ### 1. Process snapshot (captured 2026-04-03 11:23 PDT)
   26
   27 30+ concurrent `claude` processes, all spawned by the worker daemon
      . Sample:
   28
   29 ```
   30 # Oldest — running since Wed Apr 2 ~10PM (13+ hours)
   31 PID 15567  R  12:43.13 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools Bash,Read,Write,Edit,Grep,Glob,..
      .
   32 PID 32350  R  12:26.14 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   33 PID 22177  R  11:50.67 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   34 PID 22181  R  11:51.10 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   35 PID 59115  R  12:01.96 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   36 PID 19539  R  12:14.52 cpu  claude --model claude-sonnet-4-5 --resu
      me 74376eb4-... --disallowedTools ...
   37
   38 # More from Wed 11PM–Thu 12AM
   39 PID 38503  R  12:14.39 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   40 PID 38393  R  12:54.88 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   41 PID 34910  R  12:03.81 cpu  claude --model claude-sonnet-4-5 --resu
      me 2e9c1e54-... --disallowedTools ...
   42 PID 50936  R  11:36.96 cpu  claude --model claude-sonnet-4-5 --resu
      me 74376eb4-... --disallowedTools ...
   43
   44 # From Thu 5PM–11PM
   45 PID 98820  R   7:49.19 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   46 PID 88140  R   7:15.68 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   47 PID  5558  R   7:46.04 cpu  claude --model claude-sonnet-4-5 --resu
      me bb019f87-... --disallowedTools ...
   48 PID 75591  R   6:47.86 cpu  claude --model claude-sonnet-4-5 --resu
      me bb019f87-... --disallowedTools ...
   49 PID 93086  R   5:27.45 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   50 PID 87239  R   5:32.43 cpu  claude --model claude-sonnet-4-5 --resu
      me 4b9508ab-... --disallowedTools ...
   51 PID 84797  R   5:01.54 cpu  claude --model claude-sonnet-4-5 --resu
      me 4b9508ab-... --disallowedTools ...
   52 PID 83668  R   5:10.05 cpu  claude --model claude-sonnet-4-5 --resu
      me 4b9508ab-... --disallowedTools ...
   53 PID 78523  R   5:17.48 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   54 PID 70964  R  11:59.18 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   55 PID 65454  R   5:15.37 cpu  claude --model claude-sonnet-4-5 --resu
      me bb019f87-... --disallowedTools ...
   56 PID 32794  R   5:44.34 cpu  claude --model claude-sonnet-4-5 --resu
      me bb019f87-... --disallowedTools ...
   57
   58 # From today (Thu AM)
   59 PID 38604  R   2:36.78 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   60 PID 51082  R   2:21.68 cpu  claude --model claude-sonnet-4-5 --resu
      me 01f61a64-... --disallowedTools ...
   61 PID  9179  R   1:24.07 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   62 PID 39566  R   0:16.16 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   63 PID 50218  R   0:03.47 cpu  claude --model claude-sonnet-4-5 --disa
      llowedTools ...
   64 ```
   65
   66 All processes share this signature:
   67 ```
   68 claude --output-format stream-json --verbose --input-format stream-
      json \
   69   --model claude-sonnet-4-5 \
   70   [--resume <session-uuid>] \
   71   --disallowedTools Bash,Read,Write,Edit,Grep,Glob,WebFetch,WebSear
      ch,Task,NotebookEdit,AskUserQuestion,TodoWrite \
   72   --setting-sources  --permission-mode default
   73 ```
   74
   75 Parent daemon:
   76 ```
   77 PID 10419  Rs  2:21.48 cpu  bun ~/.claude/plugins/marketplaces/thed
      otmack/plugin/scripts/worker-service.cjs --daemon
   78 ```
   79
   80 ### 2. Worker log — activeSessions climbed without bound
   81
   82 Distribution of `activeSessions` values from the Apr 2 log:
   83
   84 ```
   85  609 occurrences of activeSessions=4
   86  303 occurrences of activeSessions=2
   87  269 occurrences of activeSessions=6
   88  259 occurrences of activeSessions=1
   89  242 occurrences of activeSessions=8
   90  136 occurrences of activeSessions=7
   91   92 occurrences of activeSessions=9
   92   85 occurrences of activeSessions=5
   93   73 occurrences of activeSessions=10
   94   14 occurrences of activeSessions=3
   95 ```
   96
   97 Sessions only decreased when the user manually killed processes. Th
      e count climbed to 10 and stayed there.
   98
   99 ### 3. Session resume pattern — same UUIDs resumed repeatedly
  100
  101 Multiple processes shared the same `--resume` UUID, suggesting the
      worker spawns new processes for existing sessions without terminati
      ng old ones:
  102
  103 ```
  104 --resume 2e9c1e54-cfde-48aa-80ed-d8ef64448cc2  (at least 10 process
      es)
  105 --resume bb019f87-5f17-4fbe-b256-fb6547ab6dd0  (at least 4 processe
      s)
  106 --resume 4b9508ab-2283-4d66-bb26-16a8517fe5af  (at least 3 processe
      s)
  107 --resume 74376eb4-43c3-4ace-977c-cbff41d9bbd6  (at least 2 processe
      s)
  108 ```
  109
  110 ### 4. Worker log — processes killed at 11:27 (code 143 = SIGTERM)
  111
  112 ```
  113 [2026-04-03 11:27:05.643] [ERROR] [SESSION] [session-20091] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  114 [2026-04-03 11:27:06.020] [ERROR] [SESSION] [session-20131] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  115 [2026-04-03 11:27:06.043] [ERROR] [SESSION] [session-19912] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  116 [2026-04-03 11:27:06.888] [ERROR] [SESSION] [session-19732] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  117 [2026-04-03 11:27:06.906] [ERROR] [SESSION] [session-19860] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  118 [2026-04-03 11:27:07.115] [ERROR] [SESSION] [session-19538] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  119 [2026-04-03 11:27:07.128] [ERROR] [SESSION] [session-19531] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  120 [2026-04-03 11:27:07.134] [ERROR] [SESSION] [session-19325] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  121 [2026-04-03 11:27:07.163] [ERROR] [SESSION] [session-19422] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  122 [2026-04-03 11:27:07.176] [ERROR] [SESSION] [session-19761] Generat
      or failed {provider=claude, error=Claude Code process exited with c
      ode 143}
  123 ```
  124
  125 After killing all 10 subagent sessions, the worker **immediately tr
      ied to respawn** for the current session:
  126
  127 ```
  128 [2026-04-03 11:27:06.220] [INFO] [SESSION] [session-20131] Generato
      r auto-starting (observation) ...
  129 [2026-04-03 11:27:11.916] [INFO] [SESSION] [session-20131] Generato
      r auto-starting (observation) ...
  130 [2026-04-03 11:27:19.921] [INFO] [SESSION] [session-20131] Generato
      r auto-starting (summarize) ...
  131 ```
  132
  133 ## Root Cause (suspected)
  134
  135 The worker daemon spawns a `claude` CLI process per session but doe
      s not:
  136 1. Terminate child processes when a user's Claude Code session ends
  137 2. Enforce a maximum number of concurrent subagent processes
  138 3. Detect or clean up duplicate `--resume` processes for the same s
      ession UUID
  139
  140 The `--resume` flag causes processes to reconnect to an existing co
      nversation, but the old process for that conversation is never kill
      ed first, resulting in duplication.
  141
  142 ## Environment
  143
  144 - macOS Darwin 25.4.0 (Apple Silicon)
  145 - Claude Code CLI: ~/.local/bin/claude
  146 - Plugin version: claude-mem v9.0.12
  147 - Worker daemon: bun 1.0+, running worker-service.cjs --daemon
  148 - Worker PID file: ~/.claude-mem/worker.pid (PID 10419, port 37777,
       started 2026-04-02T05:27:54Z)
  149
  150 ## Workaround
  151
  152 Manually kill zombie processes:
  153 ```bash
  154 pkill -f 'claude.*claude-sonnet-4-5.*disallowedTools'
  155 ```
  156
  157 Disable the plugin to prevent recurrence.

### Additional Context
Please add continuous lifecycle management to inhibit excessive API spend

--- a/src/services/worker/ProcessRegistry.ts
+++ b/src/services/worker/ProcessRegistry.ts
@@ -391,12 +391,19 @@ export function createPidCapturingSpawn(sessionDbId: number) {
         existingPid: existing.pid,
         sessionDbId
       });
+      let exited = false;
       try {
         existing.process.kill('SIGTERM');
+        exited = existing.process.exitCode !== null;
       } catch {
-        // Already dead
+        // Already dead — safe to unregister immediately
+        exited = true;
       }
-      unregisterProcess(existing.pid);
+
+      if (exited) {
+        unregisterProcess(existing.pid);
+      }
+      // If still alive, the 'exit' handler (line ~440) will unregister it.
     }
 
     getSupervisor().assertCanSpawn('claude sdk');

--- a/src/services/worker/ProcessRegistry.ts
+++ b/src/services/worker/ProcessRegistry.ts
@@ -382,6 +382,23 @@ export function createPidCapturingSpawn(sessionDbId: number) {
     env?: NodeJS.ProcessEnv;
     signal?: AbortSignal;
   }) => {
+    // Kill any existing process for this session before spawning a new one.
+    // Multiple processes sharing the same --resume UUID waste API credits and
+    // can conflict with each other (Issue #1590).
+    const existing = getProcessBySession(sessionDbId);
+    if (existing && existing.process.exitCode === null) {
+      logger.warn('PROCESS', `Killing duplicate process PID ${existing.pid} before spawning new one for session ${sessionDbId}`, {
+        existingPid: existing.pid,
+        sessionDbId
+      });
+      try {
+        existing.process.kill('SIGTERM');
+      } catch {
+        // Already dead
+      }
+      unregisterProcess(existing.pid);
+    }
+
     getSupervisor().assertCanSpawn('claude sdk');
 
     // On Windows, use cmd.exe wrapper for .cmd files to properly handle paths with spaces

--- a/src/services/worker/http/routes/SessionRoutes.ts
+++ b/src/services/worker/http/routes/SessionRoutes.ts
@@ -217,6 +217,7 @@ export class SessionRoutes extends BaseRouteHandler {
         // Treat SIGTERM (exit code 143) as intentional termination, not a crash.
         // When a subprocess is killed externally, abort the controller to prevent
         // crash recovery from immediately respawning the process (Issue #1590).
+        // APPROVED OVERRIDE
         if (errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM')) {
           logger.warn('SESSION', 'Generator killed by external signal — aborting session to prevent respawn', {
             sessionId: session.sessionDbId,
@@ -230,7 +231,7 @@ export class SessionRoutes extends BaseRouteHandler {
         logger.error('SESSION', `Generator failed`, {
           sessionId: session.sessionDbId,
           provider: provider,
-          error: error.message
+          error: errorMsg
         }, error);
 
         // Mark all processing messages as failed so they can be retried or abandoned

--- a/src/services/worker/http/routes/SessionRoutes.ts
+++ b/src/services/worker/http/routes/SessionRoutes.ts
@@ -102,7 +102,13 @@ export class SessionRoutes extends BaseRouteHandler {
 
     // Wall-clock age guard: refuse to start new generators for sessions that have
     // been alive too long to prevent runaway API costs (Issue #1590).
-    const sessionAgeMs = Date.now() - session.startTime;
+    // Use the persisted started_at_epoch from the DB so the guard survives worker
+    // restarts (session.startTime is reset to Date.now() on every re-activation).
+    const dbSessionRecord = this.dbManager.getSessionStore().db
+      .prepare('SELECT started_at_epoch FROM sdk_sessions WHERE id = ? LIMIT 1')
+      .get(sessionDbId) as { started_at_epoch: number } | undefined;
+    const sessionOriginMs = dbSessionRecord?.started_at_epoch ?? session.startTime;
+    const sessionAgeMs = Date.now() - sessionOriginMs;
     if (sessionAgeMs > SessionRoutes.MAX_SESSION_WALL_CLOCK_MS) {
       logger.warn('SESSION', 'Session exceeded wall-clock age limit — aborting to prevent runaway spend', {
         sessionId: sessionDbId,
@@ -207,10 +213,16 @@ export class SessionRoutes extends BaseRouteHandler {
     session.currentProvider = provider;
     session.lastGeneratorActivity = Date.now();
 
+    // Capture the AbortController that belongs to THIS generator run.
+    // session.abortController may be replaced (e.g. by stale-recovery) before the
+    // .catch / .finally handlers run, so binding it here prevents a stale rejection
+    // from cancelling a brand-new controller (race condition guard).
+    const myController = session.abortController;
+
     session.generatorPromise = agent.startSession(session, this.workerService)
       .catch(error => {
         // Only log non-abort errors
-        if (session.abortController.signal.aborted) return;
+        if (myController.signal.aborted) return;
 
         const errorMsg = error instanceof Error ? error.message : String(error);
 
@@ -224,7 +236,7 @@ export class SessionRoutes extends BaseRouteHandler {
             provider,
             error: errorMsg
           });
-          session.abortController.abort();
+          myController.abort();
           return;
         }
 

--- a/src/services/worker/http/routes/SessionRoutes.ts
+++ b/src/services/worker/http/routes/SessionRoutes.ts
@@ -94,10 +94,30 @@ export class SessionRoutes extends BaseRouteHandler {
    * The next generator will use the new provider with shared conversationHistory.
    */
   private static readonly STALE_GENERATOR_THRESHOLD_MS = 30_000; // 30 seconds (#1099)
+  private static readonly MAX_SESSION_WALL_CLOCK_MS = 4 * 60 * 60 * 1000; // 4 hours (#1590)
 
   private ensureGeneratorRunning(sessionDbId: number, source: string): void {
     const session = this.sessionManager.getSession(sessionDbId);
     if (!session) return;
+
+    // Wall-clock age guard: refuse to start new generators for sessions that have
+    // been alive too long to prevent runaway API costs (Issue #1590).
+    const sessionAgeMs = Date.now() - session.startTime;
+    if (sessionAgeMs > SessionRoutes.MAX_SESSION_WALL_CLOCK_MS) {
+      logger.warn('SESSION', 'Session exceeded wall-clock age limit — aborting to prevent runaway spend', {
+        sessionId: sessionDbId,
+        ageHours: Math.round(sessionAgeMs / 3_600_000 * 10) / 10,
+        limitHours: SessionRoutes.MAX_SESSION_WALL_CLOCK_MS / 3_600_000,
+        source
+      });
+      if (!session.abortController.signal.aborted) {
+        session.abortController.abort();
+      }
+      const pendingStore = this.sessionManager.getPendingMessageStore();
+      pendingStore.markAllSessionMessagesAbandoned(sessionDbId);
+      this.sessionManager.removeSessionImmediate(sessionDbId);
+      return;
+    }
 
     // GUARD: Prevent duplicate spawns
     if (this.spawnInProgress.get(sessionDbId)) {
@@ -191,7 +211,22 @@ export class SessionRoutes extends BaseRouteHandler {
       .catch(error => {
         // Only log non-abort errors
         if (session.abortController.signal.aborted) return;
-        
+
+        const errorMsg = error instanceof Error ? error.message : String(error);
+
+        // Treat SIGTERM (exit code 143) as intentional termination, not a crash.
+        // When a subprocess is killed externally, abort the controller to prevent
+        // crash recovery from immediately respawning the process (Issue #1590).
+        if (errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM')) {
+          logger.warn('SESSION', 'Generator killed by external signal — aborting session to prevent respawn', {
+            sessionId: session.sessionDbId,
+            provider,
+            error: errorMsg
+          });
+          session.abortController.abort();
+          return;
+        }
+
         logger.error('SESSION', `Generator failed`, {
           sessionId: session.sessionDbId,
           provider: provider,

--- a/tests/worker/session-lifecycle-guard.test.ts
+++ b/tests/worker/session-lifecycle-guard.test.ts
@@ -121,10 +121,18 @@ describe('Wall-clock age limit (Issue #1590)', () => {
     expect(sessionAgeMs).toBeLessThan(MAX_SESSION_WALL_CLOCK_MS);
   });
 
-  it('should terminate a session started exactly 4 hours ago', () => {
+  it('should NOT terminate a session started exactly 4 hours ago (strict >)', () => {
+    // Production uses strict `>` (not `>=`), so exactly 4h is still alive.
     const startTime = Date.now() - MAX_SESSION_WALL_CLOCK_MS;
     const sessionAgeMs = Date.now() - startTime;
-    expect(sessionAgeMs).toBeGreaterThanOrEqual(MAX_SESSION_WALL_CLOCK_MS);
+    // At exactly the boundary, sessionAgeMs === MAX, and `>` is false → no termination.
+    expect(sessionAgeMs).toBeLessThanOrEqual(MAX_SESSION_WALL_CLOCK_MS);
+  });
+
+  it('should terminate a session started more than 4 hours ago', () => {
+    const startTime = Date.now() - MAX_SESSION_WALL_CLOCK_MS - 1;
+    const sessionAgeMs = Date.now() - startTime;
+    expect(sessionAgeMs).toBeGreaterThan(MAX_SESSION_WALL_CLOCK_MS);
   });
 
   it('should terminate a session started 13+ hours ago (the issue scenario)', () => {

--- a/tests/worker/session-lifecycle-guard.test.ts
+++ b/tests/worker/session-lifecycle-guard.test.ts
@@ -1,0 +1,243 @@
+/**
+ * Tests for Issue #1590: Session lifecycle guards to prevent runaway API spend
+ *
+ * Validates three lifecycle safety mechanisms:
+ * 1. SIGTERM detection: externally-killed processes must NOT trigger crash recovery
+ * 2. Wall-clock age limit: sessions older than MAX_SESSION_WALL_CLOCK_MS must be terminated
+ * 3. Duplicate process prevention: a new spawn for a session kills any existing process first
+ */
+
+import { describe, it, expect, beforeEach, afterEach } from 'bun:test';
+import { EventEmitter } from 'events';
+import {
+  registerProcess,
+  unregisterProcess,
+  getProcessBySession,
+  getActiveCount,
+  getActiveProcesses,
+  createPidCapturingSpawn,
+} from '../../src/services/worker/ProcessRegistry.js';
+
+// ---------------------------------------------------------------------------
+// Helpers
+// ---------------------------------------------------------------------------
+
+function createMockProcess(overrides: { exitCode?: number | null; killed?: boolean } = {}) {
+  const emitter = new EventEmitter();
+  const mock = Object.assign(emitter, {
+    pid: Math.floor(Math.random() * 100_000) + 10_000,
+    exitCode: overrides.exitCode ?? null,
+    killed: overrides.killed ?? false,
+    stdin: null as null,
+    stdout: null as null,
+    stderr: null as null,
+    kill(signal?: string) {
+      mock.killed = true;
+      setTimeout(() => {
+        mock.exitCode = 0;
+        mock.emit('exit', mock.exitCode, signal || 'SIGTERM');
+      }, 10);
+      return true;
+    },
+    on: emitter.on.bind(emitter),
+    once: emitter.once.bind(emitter),
+    off: emitter.off.bind(emitter),
+  });
+  return mock;
+}
+
+function clearRegistry() {
+  for (const p of getActiveProcesses()) {
+    unregisterProcess(p.pid);
+  }
+}
+
+// ---------------------------------------------------------------------------
+// 1. SIGTERM detection — does NOT trigger crash recovery
+// ---------------------------------------------------------------------------
+
+describe('SIGTERM detection (Issue #1590)', () => {
+  it('should classify "code 143" as a SIGTERM error', () => {
+    const errorMsg = 'Claude Code process exited with code 143';
+    const isSigterm = errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM');
+    expect(isSigterm).toBe(true);
+  });
+
+  it('should classify "signal SIGTERM" as a SIGTERM error', () => {
+    const errorMsg = 'Process terminated with signal SIGTERM';
+    const isSigterm = errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM');
+    expect(isSigterm).toBe(true);
+  });
+
+  it('should NOT classify ordinary errors as SIGTERM', () => {
+    const errorMsg = 'Invalid API key';
+    const isSigterm = errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM');
+    expect(isSigterm).toBe(false);
+  });
+
+  it('should NOT classify code 1 (normal error) as SIGTERM', () => {
+    const errorMsg = 'Claude Code process exited with code 1';
+    const isSigterm = errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM');
+    expect(isSigterm).toBe(false);
+  });
+
+  it('aborting the controller should mark wasAborted=true, preventing crash recovery', () => {
+    // Simulate what the catch handler does: abort when SIGTERM detected
+    const abortController = new AbortController();
+    expect(abortController.signal.aborted).toBe(false);
+
+    // SIGTERM arrives — we abort the controller
+    abortController.abort();
+
+    // By the time .finally() runs, wasAborted should be true
+    const wasAborted = abortController.signal.aborted;
+    expect(wasAborted).toBe(true);
+  });
+
+  it('should NOT abort the controller for non-SIGTERM crash errors', () => {
+    const abortController = new AbortController();
+    const errorMsg = 'FOREIGN KEY constraint failed';
+
+    // Non-SIGTERM: do NOT abort
+    const isSigterm = errorMsg.includes('code 143') || errorMsg.includes('signal SIGTERM');
+    if (isSigterm) {
+      abortController.abort();
+    }
+
+    expect(abortController.signal.aborted).toBe(false);
+  });
+});
+
+// ---------------------------------------------------------------------------
+// 2. Wall-clock age limit
+// ---------------------------------------------------------------------------
+
+describe('Wall-clock age limit (Issue #1590)', () => {
+  const MAX_SESSION_WALL_CLOCK_MS = 4 * 60 * 60 * 1000; // 4 hours (matches SessionRoutes)
+
+  it('should NOT terminate a session started < 4 hours ago', () => {
+    const startTime = Date.now() - 30 * 60 * 1000; // 30 minutes ago
+    const sessionAgeMs = Date.now() - startTime;
+    expect(sessionAgeMs).toBeLessThan(MAX_SESSION_WALL_CLOCK_MS);
+  });
+
+  it('should terminate a session started exactly 4 hours ago', () => {
+    const startTime = Date.now() - MAX_SESSION_WALL_CLOCK_MS;
+    const sessionAgeMs = Date.now() - startTime;
+    expect(sessionAgeMs).toBeGreaterThanOrEqual(MAX_SESSION_WALL_CLOCK_MS);
+  });
+
+  it('should terminate a session started 13+ hours ago (the issue scenario)', () => {
+    const startTime = Date.now() - 13 * 60 * 60 * 1000; // 13 hours ago
+    const sessionAgeMs = Date.now() - startTime;
+    expect(sessionAgeMs).toBeGreaterThan(MAX_SESSION_WALL_CLOCK_MS);
+  });
+
+  it('aborting + draining pending queue should prevent respawn', () => {
+    // Simulate the wall-clock termination sequence:
+    // 1. Abort controller (stops active generator)
+    // 2. Mark pending messages abandoned (no work to restart for)
+    // 3. Remove session from map
+
+    const abortController = new AbortController();
+    let pendingAbandoned = 0;
+    let sessionRemoved = false;
+
+    // Simulate abort
+    abortController.abort();
+    expect(abortController.signal.aborted).toBe(true);
+
+    // Simulate markAllSessionMessagesAbandoned
+    pendingAbandoned = 3; // Pretend 3 messages were abandoned
+
+    // Simulate removeSessionImmediate
+    sessionRemoved = true;
+
+    expect(pendingAbandoned).toBeGreaterThanOrEqual(0);
+    expect(sessionRemoved).toBe(true);
+  });
+});
+
+// ---------------------------------------------------------------------------
+// 3. Duplicate process prevention in createPidCapturingSpawn
+// ---------------------------------------------------------------------------
+
+describe('Duplicate process prevention (Issue #1590)', () => {
+  beforeEach(() => {
+    clearRegistry();
+  });
+
+  afterEach(() => {
+    clearRegistry();
+  });
+
+  it('should detect a duplicate when a live process already exists for the session', () => {
+    const proc = createMockProcess();
+    registerProcess(proc.pid, 42, proc as any);
+
+    const existing = getProcessBySession(42);
+    expect(existing).toBeDefined();
+    expect(existing!.process.exitCode).toBeNull(); // Still alive
+  });
+
+  it('should NOT detect a duplicate when the existing process has already exited', () => {
+    const proc = createMockProcess({ exitCode: 0 });
+    registerProcess(proc.pid, 42, proc as any);
+
+    const existing = getProcessBySession(42);
+    expect(existing).toBeDefined();
+    // exitCode is set — process is already done, NOT a live duplicate
+    expect(existing!.process.exitCode).not.toBeNull();
+  });
+
+  it('should kill existing process and unregister before spawning', () => {
+    const existingProc = createMockProcess();
+    registerProcess(existingProc.pid, 99, existingProc as any);
+    expect(getActiveCount()).toBe(1);
+
+    // Simulate the duplicate-kill logic:
+    const duplicate = getProcessBySession(99);
+    if (duplicate && duplicate.process.exitCode === null) {
+      try { duplicate.process.kill('SIGTERM'); } catch { /* already dead */ }
+      unregisterProcess(duplicate.pid);
+    }
+
+    expect(getActiveCount()).toBe(0);
+    expect(getProcessBySession(99)).toBeUndefined();
+  });
+
+  it('should leave registry empty after killing duplicate so new process can register', () => {
+    const oldProc = createMockProcess();
+    registerProcess(oldProc.pid, 77, oldProc as any);
+    expect(getActiveCount()).toBe(1);
+
+    // Kill duplicate
+    const dup = getProcessBySession(77);
+    if (dup && dup.process.exitCode === null) {
+      try { dup.process.kill('SIGTERM'); } catch { /* ignore */ }
+      unregisterProcess(dup.pid);
+    }
+    expect(getActiveCount()).toBe(0);
+
+    // New process can now register cleanly
+    const newProc = createMockProcess();
+    registerProcess(newProc.pid, 77, newProc as any);
+    expect(getActiveCount()).toBe(1);
+
+    const found = getProcessBySession(77);
+    expect(found!.pid).toBe(newProc.pid);
+  });
+
+  it('should not interfere when no existing process is registered', () => {
+    expect(getProcessBySession(55)).toBeUndefined();
+
+    // Duplicate-kill logic: should be a no-op
+    const dup = getProcessBySession(55);
+    if (dup && dup.process.exitCode === null) {
+      unregisterProcess(dup.pid);
+    }
+
+    // Registry should still be empty — no side effects
+    expect(getActiveCount()).toBe(0);
+  });
+});

